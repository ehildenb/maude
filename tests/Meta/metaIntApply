#!/bin/sh

MAUDE_LIB=$srcdir/../../src/Main
export MAUDE_LIB

../../src/Main/maude \
  < $srcdir/metaIntApply.maude -no-banner -no-advise \
  > metaIntApply.out 2>&1

diff $srcdir/metaIntApply.expected metaIntApply.out > /dev/null 2>&1
