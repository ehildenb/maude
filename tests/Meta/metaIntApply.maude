set show timing off .
set show advisories off .

***
*** Test applyRule() in the meta-interpreter without and with extension.
***

load metaInterpreter

mod APPLY is
  sorts Foo .
  ops a b : -> Foo .
  op f : Foo Foo -> Foo [comm] .

  rl f(X:Foo, Y:Foo) => X:Foo [label k] .
endm

mod APPLY-TEST is
  pr META-INTERPRETER .

  op me : -> Oid .
  op User : -> Cid .
  op soln:_ : Nat -> Attribute .
  op got:_  : Msg -> Attribute .
  op subst:_ : Substitution -> Attribute .

  vars X Y Z : Oid .
  var AS : AttributeSet .
  vars N M : Nat .
  vars IS S : Substitution .
  var T : Term .
  var TY : Type .

  rl < X : User | AS > createdInterpreter(X, Y, Z) =>
     < X : User | AS > insertModule(Z, X, upModule('APPLY, true)) .

  rl < X : User | AS, subst: IS > insertedModule(X, Y) =>
     < X : User | AS, soln: 1, subst: IS > applyRule(Y, X, 'APPLY, 'f['a.Foo, 'b.Foo], 'k, IS, 0) .

  rl < X : User | AS, soln: N, subst: IS > appliedRule(X, Y, M, T, TY, S) =>
     < X : User | AS, soln: (N + 1),  subst: IS, got: appliedRule(X, Y, M, T, TY, S) >
     applyRule(Y, X, 'APPLY, 'f['a.Foo, 'b.Foo], 'k, IS, N) .
endm

set show breakdown on .

erew in APPLY-TEST : <> 
< me : User | subst: none > 
createInterpreter(interpreterManager, me, none) .

erew in APPLY-TEST : <> 
< me : User | subst: ('Y:Foo <- 'b.Foo) > 
createInterpreter(interpreterManager, me, none) .


mod APPLY2 is
  sorts Foo .
  ops a b c  : -> Foo .
  op f : Foo Foo -> Foo [assoc comm] .
  op g : Foo -> Foo .

  rl f(X:Foo, Y:Foo) => X:Foo [label k] .
endm

mod APPLY-TEST2 is
  pr META-INTERPRETER .

  op me : -> Oid .
  op User : -> Cid .
  op soln:_ : Nat -> Attribute .
  op got:_  : Msg -> Attribute .
  op subst:_ : Substitution -> Attribute .

  vars X Y Z : Oid .
  var AS : AttributeSet .
  vars N M : Nat .
  vars IS S : Substitution .
  var T : Term .
  var TY : Type .
  var C : Context .

  rl < X : User | AS > createdInterpreter(X, Y, Z) =>
     < X : User | AS > insertModule(Z, X, upModule('APPLY2, true)) .

  rl < X : User | AS, subst: IS > insertedModule(X, Y) =>
     < X : User | AS, soln: 1, subst: IS > applyRule(Y, X, 'APPLY2, 'g['g['f['a.Foo, 'b.Foo, 'c.Foo]]], 'k, IS, 0, unbounded, 0) .

  rl < X : User | AS, soln: N, subst: IS > appliedRule(X, Y, M, T, TY, S, C) =>
     < X : User | AS, soln: (N + 1),  subst: IS, got: appliedRule(X, Y, M, T, TY, S, C) >
     applyRule(Y, X, 'APPLY2, 'g['g['f['a.Foo, 'b.Foo, 'c.Foo]]], 'k, IS, 0, unbounded, N) .
endm

erew in APPLY-TEST2 : <> 
< me : User | subst: none > 
createInterpreter(interpreterManager, me, none) .

erew in APPLY-TEST2 : <> 
< me : User | subst: ('X:Foo <- 'a.Foo) > 
createInterpreter(interpreterManager, me, none) .
