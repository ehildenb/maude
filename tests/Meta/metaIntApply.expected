==========================================
erewrite in APPLY-TEST : <> < me : User | subst: none > createInterpreter(
    interpreterManager, me, none) .
rewrites: 9
mb applications: 0  equational rewrites: 3  rule rewrites: 6  variant narrowing
    steps: 0  narrowing steps: 0
result Configuration: <> < me : User | soln: 3,got: appliedRule(me,
    interpreter(0), 1, 'a.Foo, 'Foo, 
  'X:Foo <- 'a.Foo ; 
  'Y:Foo <- 'b.Foo),got: appliedRule(me, interpreter(0), 1, 'b.Foo, 'Foo, 
  'X:Foo <- 'b.Foo ; 
  'Y:Foo <- 'a.Foo),subst: none > noSuchResult(me, interpreter(0), 0)
==========================================
erewrite in APPLY-TEST : <> < me : User | subst: (
  'Y:Foo <- 'b.Foo) > createInterpreter(interpreterManager, me, none) .
rewrites: 6
mb applications: 0  equational rewrites: 2  rule rewrites: 4  variant narrowing
    steps: 0  narrowing steps: 0
result Configuration: <> < me : User | soln: 2,got: appliedRule(me,
    interpreter(0), 1, 'a.Foo, 'Foo, 
  'X:Foo <- 'a.Foo ; 
  'Y:Foo <- 'b.Foo),subst: (
  'Y:Foo <- 'b.Foo) > noSuchResult(me, interpreter(0), 0)
==========================================
erewrite in APPLY-TEST2 : <> < me : User | subst: none > createInterpreter(
    interpreterManager, me, none) .
rewrites: 39
mb applications: 0  equational rewrites: 13  rule rewrites: 26  variant
    narrowing steps: 0  narrowing steps: 0
result Configuration: <> < me : User | soln: 13,got: appliedRule(me,
    interpreter(0), 1, 'g['g['a.Foo]], 'Foo, 
  'X:Foo <- 'a.Foo ; 
  'Y:Foo <- 'f['b.Foo,'c.Foo], 'g['g[[]]]),got: appliedRule(me, interpreter(0),
    1, 'g['g['b.Foo]], 'Foo, 
  'X:Foo <- 'b.Foo ; 
  'Y:Foo <- 'f['a.Foo,'c.Foo], 'g['g[[]]]),got: appliedRule(me, interpreter(0),
    1, 'g['g['c.Foo]], 'Foo, 
  'X:Foo <- 'c.Foo ; 
  'Y:Foo <- 'f['a.Foo,'b.Foo], 'g['g[[]]]),got: appliedRule(me, interpreter(0),
    1, 'g['g['f['a.Foo,'b.Foo]]], 'Foo, 
  'X:Foo <- 'a.Foo ; 
  'Y:Foo <- 'c.Foo, 'g['g['f['b.Foo,[]]]]),got: appliedRule(me, interpreter(0),
    1, 'g['g['f['a.Foo,'b.Foo]]], 'Foo, 
  'X:Foo <- 'b.Foo ; 
  'Y:Foo <- 'c.Foo, 'g['g['f['a.Foo,[]]]]),got: appliedRule(me, interpreter(0),
    1, 'g['g['f['a.Foo,'b.Foo]]], 'Foo, 
  'X:Foo <- 'f['a.Foo,'b.Foo] ; 
  'Y:Foo <- 'c.Foo, 'g['g[[]]]),got: appliedRule(me, interpreter(0), 1, 'g['g[
    'f['a.Foo,'c.Foo]]], 'Foo, 
  'X:Foo <- 'a.Foo ; 
  'Y:Foo <- 'b.Foo, 'g['g['f['c.Foo,[]]]]),got: appliedRule(me, interpreter(0),
    1, 'g['g['f['a.Foo,'c.Foo]]], 'Foo, 
  'X:Foo <- 'c.Foo ; 
  'Y:Foo <- 'b.Foo, 'g['g['f['a.Foo,[]]]]),got: appliedRule(me, interpreter(0),
    1, 'g['g['f['a.Foo,'c.Foo]]], 'Foo, 
  'X:Foo <- 'f['a.Foo,'c.Foo] ; 
  'Y:Foo <- 'b.Foo, 'g['g[[]]]),got: appliedRule(me, interpreter(0), 1, 'g['g[
    'f['b.Foo,'c.Foo]]], 'Foo, 
  'X:Foo <- 'b.Foo ; 
  'Y:Foo <- 'a.Foo, 'g['g['f['c.Foo,[]]]]),got: appliedRule(me, interpreter(0),
    1, 'g['g['f['b.Foo,'c.Foo]]], 'Foo, 
  'X:Foo <- 'c.Foo ; 
  'Y:Foo <- 'a.Foo, 'g['g['f['b.Foo,[]]]]),got: appliedRule(me, interpreter(0),
    1, 'g['g['f['b.Foo,'c.Foo]]], 'Foo, 
  'X:Foo <- 'f['b.Foo,'c.Foo] ; 
  'Y:Foo <- 'a.Foo, 'g['g[[]]]),subst: none > noSuchResult(me, interpreter(0),
    0)
==========================================
erewrite in APPLY-TEST2 : <> < me : User | subst: (
  'X:Foo <- 'a.Foo) > createInterpreter(interpreterManager, me, none) .
rewrites: 12
mb applications: 0  equational rewrites: 4  rule rewrites: 8  variant narrowing
    steps: 0  narrowing steps: 0
result Configuration: <> < me : User | soln: 4,got: appliedRule(me,
    interpreter(0), 1, 'g['g['a.Foo]], 'Foo, 
  'X:Foo <- 'a.Foo ; 
  'Y:Foo <- 'f['b.Foo,'c.Foo], 'g['g[[]]]),got: appliedRule(me, interpreter(0),
    1, 'g['g['f['a.Foo,'b.Foo]]], 'Foo, 
  'X:Foo <- 'a.Foo ; 
  'Y:Foo <- 'c.Foo, 'g['g['f['b.Foo,[]]]]),got: appliedRule(me, interpreter(0),
    1, 'g['g['f['a.Foo,'c.Foo]]], 'Foo, 
  'X:Foo <- 'a.Foo ; 
  'Y:Foo <- 'b.Foo, 'g['g['f['c.Foo,[]]]]),subst: (
  'X:Foo <- 'a.Foo) > noSuchResult(me, interpreter(0), 0)
Bye.
